tensorrt_llm/_torch/speculative/save_hidden_state.py

```diff
@@ -0,0 +1,143 @@
+        self.last_hidden_states = None
+        if config.save_last_layer_post_norm:
+            self.last_hidden_states = torch.empty(
+                (max_num_tokens, self.hidden_size),
+                dtype=self.dtype,
+                device='cuda')
+
+
+class SaveHiddenStatesDrafter(Drafter):
+
+    def __init__(
+        self,
+        spec_config: "SaveHiddenStatesDecodingConfig",
+        spec_resource_manager: SaveHiddenStatesResourceManager,
+    ):
+        super().__init__(spec_config.max_concurrency)
+        self.spec_config = spec_config
+        self.max_draft_len = spec_config.max_draft_len
+        self._iter = 1
+        self._output_directory = spec_config.output_directory
+        self._file_prefix = spec_config.file_prefix
+        self._write_interval = spec_config.write_interval
+        self._saved_state = []
+        self.spec_resource_manager = spec_resource_manager
+        os.makedirs(self._output_directory, exist_ok=True)
+
+    def _process_request(
+            self, request: LlmRequest,
+            resource_manager: SaveHiddenStatesResourceManager) -> None:
+        out_dict = {}
+        if local_mpi_rank() == 0:
+            input_ids = torch.tensor(list(request.get_tokens(0)),
+                                     dtype=torch.long,
+                                     device='cpu')
+            hidden_size = resource_manager.hidden_size
+            num_tokens = input_ids.shape[0]
+            if self.spec_config.save_last_layer_post_norm:
+                hidden_states = resource_manager.last_hidden_states[:
+                                                                    num_tokens, :].cpu(
+                                                                    ).clone()
+            else:
+                hidden_states = resource_manager.hidden_states[:num_tokens,
+                                                               -hidden_size:].cpu(
+                                                               ).clone()
+
+            out_dict = {
+                "id":
+                self._iter,
+                "input_ids":
+                input_ids,
+                "hidden_state_features":
+                resource_manager.hidden_states[:num_tokens, :].cpu().clone(),
+                "hidden_state":
+                hidden_states,
+            }
+
+            self._saved_state.append(out_dict)
+
+    def _write_to_file(self) -> None:
+        if local_mpi_rank() == 0:
+            output_path = os.path.join(self._output_directory,
+                                       f"{self._file_prefix}_{self._iter}.pt")
+            torch.save(self._saved_state, output_path)
+        self._saved_state = []
+
+    def prepare_draft_tokens(
+        self,
+        scheduled_requests: ScheduledRequests,
+        resource_manager: Optional[ResourceManager] = None,
+    ) -> None:
+        for request in sorted(
+                scheduled_requests.context_requests,
```

Comment:
Why are we attaching `py_draft_tokens` to context requests? Pretty sure these will just get ignored, `ModelEngine` only inspects generation requests for draft tokens